phpBB/phpbb/filesystem/filesystem.php

```diff
@@ -0,0 +1,916 @@
+<?php
+/**
+ *
+ * This file is part of the phpBB Forum Software package.
+ *
+ * @copyright (c) phpBB Limited <https://www.phpbb.com>
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\filesystem;
+
+use phpbb\filesystem\exception\filesystem_exception;
+
+/**
+ * A class with various functions that are related to paths, files and the filesystem
+ */
+class filesystem implements filesystem_interface
+{
+	/**
+	 * Store some information about file ownership for phpBB's chmod function
+	 *
+	 * @var array
+	 */
+	protected $chmod_info;
+
+	/**
+	 * Stores current working directory
+	 *
+	 * @var string|bool		current working directory or false if it cannot be recovered
+	 */
+	protected $working_directory;
+
+	/**
+	 * Symfony's Filesystem component
+	 *
+	 * @var \Symfony\Component\Filesystem\Filesystem
+	 */
+	protected $symfony_filesystem;
+
+	/**
+	 * Constructor
+	 */
+	public function __construct()
+	{
+		$this->chmod_info			= array();
+		$this->symfony_filesystem	= new \Symfony\Component\Filesystem\Filesystem();
+		$this->working_directory	= null;
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function chgrp($files, $group, $recursive = false)
+	{
+		try
+		{
+			$this->symfony_filesystem->chgrp($files, $group, $recursive);
+		}
+		catch (\Symfony\Component\Filesystem\Exception\IOException $e)
+		{
+			// Try to recover filename
+			// By the time this is written that is at the end of the message
+			$error = trim($e->getMessage());
+			$file = substr($error, strrpos($error, ' '));
+
+			throw new filesystem_exception('CANNOT_CHANGE_FILE_GROUP', $file, array(), $e);
+		}
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function chmod($files, $perms = null, $recursive = false, $force_chmod_link = false)
+	{
+		if (is_null($perms))
+		{
+			// Default to read permission for compatibility reasons
+			$perms = self::CHMOD_READ;
+		}
+
+		// Check if we got a permission flag
+		if ($perms > self::CHMOD_ALL)
+		{
+			$file_perm = $perms;
+
+			// Extract permissions
+			//$owner = ($file_perm >> 6) & 7; // This will be ignored
+			$group = ($file_perm >> 3) & 7;
+			$other = ($file_perm >> 0) & 7;
+
+			// Does any permissions provided? if so we add execute bit for directories
+			$group = ($group !== 0) ? ($group | self::CHMOD_EXECUTE) : $group;
+			$other = ($other !== 0) ? ($other | self::CHMOD_EXECUTE) : $other;
```

Comment:
but actually we need the execution bit to write to the directory... so we should fix the interface instead =D